### PR TITLE
Solutions to Lecture 1

### PR DESCRIPTION
**Pull request by jkomyno**
### Solutions for Lecture 1

I have tried to keep the types as generic as possible, and to use pointfree notation when useful.

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -50,7 +51,11 @@ Explanation: @sumOfSquares 3 4@ should be equal to @9 + 16@ and this
 is 25.
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-sumOfSquares x y = error "TODO!"
+sumOfSquares :: Num a => a -> a -> a
```
Comment (chshersh) [inline src/Lecture1.hs:54]: `Num` is explained in the following lectures, so you are one step ahead 😸
In here `Int -> Int -> Int` or `Integer -> Integer -> Integer` would also work, as a more specific type 🙂

File: src/Lecture1.hs [hunk 2]
```diff
@@ -63,7 +68,10 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Integral a => a -> a
+lastDigit n
+    | n <= 0    = (-n) `mod` 10
+    | otherwise = n `mod` 10
```
Comment (chshersh) [inline src/Lecture1.hs:74]: That's correct! 💯 
You can simplify the implementation a bit using the standard function `abs` that returns an absolute value of a number:

```haskell
λ: abs 7
7
λ: abs (-7)
7
```

File: src/Lecture1.hs [hunk 3]
```diff
@@ -104,7 +119,10 @@ and finds a sum of the numbers inside this string.
 
 The string contains only spaces and/or numbers.
 -}
-strSum str = error "TODO"
+strSum :: String -> Int
+strSum = let
+    toInt = read :: String -> Int
```
Comment (chshersh) [inline src/Lecture1.hs:124]: As you specify the top-level type signature for this function, the compiler would be able to figure out what type `read` function has in this case due to type inference, so you can skip specifying it in the body.

File: src/Lecture1.hs [hunk 4]
```diff
@@ -119,4 +137,8 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 
 🕯 HINT: Use recursion to implement this function.
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: (Show a, Ord a) => a -> [a] -> String
+lowerAndGreater n list = let
+    countLower   = length $ filter (< n) list
+    countGreater = length $ filter (> n) list
```
Comment (chshersh) [inline src/Lecture1.hs:143]: That's correct! 💯 

However, with this solution, you will walk through the list twice.
With the recursive solution and manual check on every element in the list on both conditions at the same time (without `filter` function), you would be able to write it more performant with only one traversal of the list 🙂

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

File: src/Lecture1.hs (hunk 2 above)
Comment (jkomyno) [reply src/Lecture1.hs:74]: True, thanks!

**Comment by chshersh** [comment]
Hi @jkomyno 👋🏻 
I see that you've added Lecture 2 solutions to this PR. It's more convenient for me to review when you open PRs for different lectures. You can also merge ready solutions 🙂

**Pull request merged**